graphix/channels.py

```diff
@@ -2,78 +2,137 @@
+        Operator.
+    """
+
+    coef: complex
+    operator: npt.NDArray[np.complex128]
+
+    def __init__(self, coef: complex, operator: npt.NDArray[_T]) -> None:
+        self.coef = coef
+        self.operator = operator.astype(np.complex128, copy=True)
+
+    @property
+    def nqubit(self) -> int:
+        """Validate the data."""
+        if not lv.is_square(self.operator):
+            raise ValueError("Operator must be a square matrix.")
+        if not lv.is_qubitop(self.operator):
+            raise ValueError("Operator must be a qubit operator.")
+        size, _ = self.operator.shape
+        return _ilog2(size)
+
 
 class KrausChannel:
     r"""Quantum channel class in the Kraus representation.
 
     Defined by Kraus operators :math:`K_i` with scalar prefactors :code:`coef`) :math:`c_i`,
     where the channel act on density matrix as :math:`\rho'  = \sum_i K_i^\dagger \rho K_i`.
-    The data should satisfy :math:`\sum K_i^\dagger K_i = I`
+    The data should satisfy :math:`\sum K_i^\dagger K_i = I`.
+    """
 
-    Attributes
-    ----------
-    nqubit : int
-        number of qubits acted on by the Kraus operators
-    size : int
-        number of Kraus operators (== Choi rank)
-    kraus_ops : list(dict())
-        the data in format
-        list(dict): [{coef: scalar, operator: list}, {coef: scalar, operator: list}, ...]
+    __nqubit: int
+    __data: list[KrausData]
 
-    """
+    @staticmethod
+    def _nqubit(kraus_data: Iterable[KrausData]) -> int:
+        try:
+            nqubits = set(data.nqubit for data in kraus_data)
+        except ValueError as e:
+            raise ValueError("Failed to intialize KrausChannel object.") from e
+
+        if len(nqubits) != 1:
+            raise ValueError("All operators must have the same shape.")
+        (nqubit,) = nqubits
 
-    def __init__(self, kraus_data):
+        return nqubit
+
+    def __init__(self, kraus_data: Iterable[KrausData]) -> None:
         """Initialize `KrausChannel` given a Kraus operator.
 
         Parameters
         ----------
-        kraus_data : list
-            array of Kraus operator data.
-            list(dict): [{coef: scalar, operator: list}, {parameter: scalar, operator: list}, ...]
-            only works for square Kraus operators
+        kraus_data : Iterable[KrausData]
+            Iterable of Kraus operator data.
 
         Raises
         ------
         ValueError
-            If empty list is provided.
+            If kraus_data is empty.
         """
-        # check there is data
         if not kraus_data:
             raise ValueError("Cannot instantiate the channel with empty data.")
 
-        if not isinstance(kraus_data, (list, np.ndarray, tuple)):
-            raise TypeError(f"The data must be a list, a numpy.ndarray or a tuple not a {type(kraus_data)}.")
+        self.__nqubit = self._nqubit(kraus_data)
+        self.__data = list(copy.deepcopy(data) for data in kraus_data)
```

Comment:
> not sure we want to impose a deep copy here

I inserted `deepcopy` here based on https://en.wikipedia.org/wiki/Principle_of_least_astonishment .

Reply:
As a user, I would not expect a data structure that performs read-only access to duplicate data. While performing a deep copy or not may be negligible in this particular case, the question of whether to copy or not arises throughout the API. I believe the more uniform we can be, the better, particularly in terms of the principle of least astonishment. The correct rule of thumb, in my opinion, is to assume that unless explicitly documented otherwise, all data passed as arguments should be treated as passed by read-only reference, and should only be copied internally if there is a need to mutate it.

Reply:
The problem is that we cannot expect users use our public APIs correctly.
As this class is defined to be public (no `_` prefix), we should make the API as ergonomic as possible, or in other words, easy to use and hard to misuse.
Let me emphasize that I can understand what you said: As a developer, I myself always bear a similar principle as yours in mind, and in fact if this class were to be defined as module-private, I would be willing to omit `deepcopy` here.

@shinich1

How do you think about the API design?
I'm focusing on safety at the cost of additional costs whereas @thierry-martinez prioritizes runtime cost assuming that users respect docstrings/Python good practices.